plugins/org.eclipse.xtext/src/org/eclipse/xtext/naming/QualifiedName.java

```diff
@@ -102,25 +123,25 @@ public static QualifiedName createFromStream(EObjectInputStream eObjectInputStre
 			// first was null, read another string which is the actual first segment
 			firstSegment = eObjectInputStream.readSegmentedString();
 		}
-		
+
 		String[] segments = readSegmentArray(eObjectInputStream, segmentCount, firstSegment);
 		if (lowerCase) {
 			return new QualifiedNameLowerCase(segments);
 		} else {
-			return new QualifiedName(segments);	
+			return new QualifiedName(segments);
 		}
 	}
 
 	private static String[] readSegmentArray(EObjectInputStream from, int count, String first) throws IOException {
 		String[] segments = new String[count];
-		segments[0] = first;
+		segments[0] = first == null? null : intern(first);
```

Comment:
It can theoretically return null, but that would indicate that we wrote a null value into the string in the first place. Since no segment of a QN may ever be null, it would be wrong to tolerate null values here. Thus I'd prefer to raise an exception.